sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala

```diff
@@ -212,4 +212,20 @@ class InnerJoinSuite extends SparkPlanTest with SharedSQLContext {
     )
   }
 
+  {
+    lazy val left = myTestData.where("a = 1")
+    lazy val right = myTestData.where("a = 1")
+    testInnerJoin(
+      "inner join, null safe",
+      left,
+      right,
+      () => (left.col("a") <=> right.col("a")).expr,
+      Seq(
+        (1, 1, 1, 1),
+        (1, 1, 1, 2),
+        (1, 2, 1, 1),
+        (1, 2, 1, 2)
+      )
```

Comment:
Is there anyway we could include data such that it'll break if the original predicate is missing?  Maybe something like

```scala
val df = Seq((1, Some(0)), (2, None)).toDF("a", "b")
checkAnswer(
  df.as("x").join(df.as("y"), $"x.b" === $"y.b"),
  Row(1, 0, 1, 0) :: Row(2, null, 2, null) :: Nil)
```